### PR TITLE
Convenient Macro for Creating ACSets

### PR DESCRIPTION
**Pull request by olynch**
Right now this is very basic; I plan to add data validation and possibly new syntaxes later on. Currently,
``` julia
g = @acset Graph begin
  V = 2
  E = 2
  src = [1,2]
  tgt = [2,1]
end
```
works, but this will not successfully reject most ill-formed ACSets.

### CODE REVIEW

**Comment by epatters** [comment]
Thanks, it will be very convenient to have a constructor like this, especially for interactive work and unit tests. I get tired of writing `add_parts!` over and over again.

**Comment by jpfairbanks** [comment]
@epatters, does this need anything else before we can merge? Let's not hold up this initial feature for the additional features like robust input validation and the CSV style syntax

**Comment by epatters** [comment]
I agree, but the docstrings need to be cleaned up and most importantly, we need to have unit tests.

**Comment by olynch** [comment]
OK, I had some grand plans also involving some @generated code, but I guess I'll just polish this a little and ship it, and do that later.

**Comment by jpfairbanks** [comment]
Yeah, I'm doing stuff with port graphs now and this would be super handy for interactive work. I'm currently doing a lot on paper and then typing it in.

**Comment by olynch** [comment]
OK, I'd feel comfortable merging this now. Sorry for the unnecessary delay.

**Review by jpfairbanks — APPROVED** [review]
LGTM! I like that because this is a macro, it should create compile time constants if you use it like:

```julia
triangle() = @acset Graph begin
  V = 3
  E = 3
  src = [1,2,3]
  tgt = [2,3,1]
end

hexagon() = [... some construction of pushouts of triangles ...]
```

that constant should propagate through the code base.

File: src/categorical_algebra/CSetDataStructures.jl
Comment (epatters) [inline src/categorical_algebra/CSetDataStructures.jl:12]: This import looks to be unused.

File: src/categorical_algebra/CSetDataStructures.jl
Comment (epatters) [inline src/categorical_algebra/CSetDataStructures.jl:827]: This function is publicly exported, but should it be? If it is exported, it should have a real docstring.

File: src/categorical_algebra/CSetDataStructures.jl
Comment (epatters) [inline src/categorical_algebra/CSetDataStructures.jl:834]: I do like to use `@match` for stuff like this:

```julia
lhs, rhs = @match elem begin
  Expr(:(=), lhs, rhs) => (lhs, rhs)
  _ => error("Every line of `@acset` must be an assignment")
end
```

In this case it's not any shorter but IMO it's easier to read.

**Review by epatters — COMMENTED** [review]
Thanks Owen! Looks good, just a few small comments.

**Comment by olynch** [comment]
Yup, those changes all make sense!

**Comment by epatters** [comment]
Thanks, LGTM! Merging now.

**Pull request merged**
